### PR TITLE
Support for LM3S: RCC and UART

### PR DESCRIPTION
**Pull request by danielinux**
Hello, 
please consider the introduction of RCC and UART support for the LM3S platform. Even if old, this target allows to run on qemu-system-arm which is often useful.

### CODE REVIEW

**Comment by karlp** [comment]
To help test this out, do you have any nice summary of using this target with qemu for the rest of us?

File: include/libopencm3/lm3s/rcc.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+enum rcc_periph_clken { CLK_ENA_E };
+enum rcc_periph_rst { CLK_RST_E };
+
+#include <libopencm3/stm32/common/rcc_common_all.h>
+
+/* --- RCC registers ------------------------------------------------------- */
+
+#define RCC_RIS             MMIO32(0x400FE050)
+#define RCC_CR			    MMIO32(0x400FE060)
+#define RCC2_CR				MMIO32(0x400FE070)
```
Comment (karlp) [inline include/libopencm3/lm3s/rcc.h]: I believe there must be a spacing problem here.

File: include/libopencm3/lm3s/rcc.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ */
+
+#ifndef LIBOPENCM3_RCC_H
+#define LIBOPENCM3_RCC_H
+#include <libopencm3/cm3/common.h>
+
+/* Interface required types */
+
+enum rcc_periph_clken { CLK_ENA_E };
+enum rcc_periph_rst { CLK_RST_E };
```
Comment (karlp) [inline include/libopencm3/lm3s/rcc.h]: An enum of one?  At the very very least, can the values start with RCC_ so as not to totally pollute the namespace?

File: include/libopencm3/lm3s/rcc.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+/* RCC1 bits */
+#define RCC_SYSDIV    (1 << 23)
+#define RCC_SYSDIV_BITPOS  (23)
+
+#define RCC_USESYSDIV (1 << 22)
+#define RCC_USEPWMDIV (1 << 20)
+
+#define RCC_PWMDIV    (1 << 17)
+#define RCC_PWMDIV_BITPOS  (17)
```
Comment (karlp) [inline include/libopencm3/lm3s/rcc.h]: This _BITPOS style is not used anywhere else, do you have any particular reason to introduce this sort of thing?  (Applies to all the other places you use these duplicate forms)  The only places an explicit shift or mask is defined normally is for multibit fields.

File: include/libopencm3/lm3s/usart.h [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef SAM3X_USART_H
+#define SAM3X_USART_H
```
Comment (karlp) [inline include/libopencm3/lm3s/usart.h]: wat

File: include/libopencm3/lm3s/usart.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+#include <libopencm3/cm3/common.h>
+
+
+
+#define USART0		0x4000C000
+#define USART1		0x4000D000
+#define USART2		0x4000E000
```
Comment (karlp) [inline include/libopencm3/lm3s/usart.h]: traditionally, base addresses for peripherals are defined as _BASE

File: include/libopencm3/lm3s/usart.h [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	USART_STOPBITS_2,
+};
+
+enum usart_parity {
+	USART_PARITY_EVEN,
+	USART_PARITY_ODD,
+	USART_PARITY_SPACE,
+	USART_PARITY_MARK,
+	USART_PARITY_NONE,
+	USART_PARITY_MULTIDROP,
```
Comment (karlp) [inline include/libopencm3/lm3s/usart.h]: I admit, I've not read the lm3s ref manuals, but multidrop parity?!  Is this really a parity setting?

File: lib/lm3s/rcc.c [hunk 7]
```diff
@@ -0,0 +1,105 @@
+/** @defgroup rcc_file RCC Controller
+
+@brief <b>LM3S RCC Controller</b>
+
+@ingroup LM3Sxx
+
+@version 1.0.0
+
+@author @htmlonly &copy; @endhtmlonly 2011
```
Comment (karlp) [inline lib/lm3s/rcc.c]: date doesn't match

File: lib/lm3s/rcc.c [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+LGPL License Terms @ref lgpl_license
+*/
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2009 Uwe Hermann <uwe@hermann-uwe.de>
+ * Copyright (C) 2009 Federico Ruiz-Ugalde <memeruiz at gmail dot com>
+ * Copyright (C) 2011 Fergus Noble <fergusnoble@gmail.com>
+ * Copyright (C) 2011 Stephen Caudle <scaudle@doceme.com>
+ * Copyright (C) 2015 Daniele Lacamera <root@danielinux.net>
```
Comment (karlp) [inline lib/lm3s/rcc.c]: as elsewhere, I'm hard pressed to believe there's anything of _any_ relevance from any other authors here, certainly not as far back as 2009.

File: lib/lm3s/rcc.c [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "stdint.h"
```
Comment (karlp) [inline lib/lm3s/rcc.c]: std headers normally get standard < and > includes, not " and " includes.

**Comment by danielinux** [comment]
@karlp OK, sorry for the noise. Will submit a cleaner patch.

File: lib/lm3s/rcc.c [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+#include "libopencm3/lm3s/rcc.h"
```
Comment (karlp) [inline lib/lm3s/rcc.c]: as do internal includes.  Please look at the rest of the library for examples.

File: lib/lm3s/rcc.c [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+extern void __dmb(void);
```
Comment (karlp) [inline lib/lm3s/rcc.c]:
```
#include <libopencm3/cm3/sync.h>
```

No need for the extern.

File: lib/lm3s/rcc.c (hunk 8 above)
Comment (danielinux) [inline lib/lm3s/rcc.c]: Tried to keep original copyright from other modules I've based this on.

File: lib/lm3s/rcc.c [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+
+/* PLL/Systick Configured values */
+#define QEMU_SYSDIV_VAL        (3)
+#define QEMU_PWMDIV_VAL        (7)
+#define QEMU_XTAL_VAL          (14)
```
Comment (karlp) [inline lib/lm3s/rcc.c]: These feel like poor names.  I'm totally totally interested in being able to run libopencm3 code under qemu, but some magic qemu defines definitely seems out of scope.  Surely it should just be, "init to these values, which are _also_ totally normal useful values, not just for qemu, and it will work well with qemu"  I'm not really sure I see why qemu needs it's own defines here.

File: lib/lm3s/rcc.c [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+int rcc_lm3s_init(uint32_t sys_div, uint32_t pwm_div, uint32_t xtal)
+{
+    uint32_t rcc = RCC_DEFAULT_VALUE;
+    uint32_t rcc2 = RCC2_DEFAULT_VALUE;
+
+    /* Initials: set default values */
+
+    RCC_CR = rcc;
+    RCC2_CR = rcc2;
```
Comment (karlp) [inline lib/lm3s/rcc.c]: Given what you do down on line 70-71 what's the point in these two lines?

File: lib/lm3s/rcc.c (hunk 10 above)
Comment (danielinux) [inline lib/lm3s/rcc.c]: :+1:

File: lib/lm3s/rcc.c [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        (xtal << RCC_XTAL_BITPOS) | RCC_USEPWMDIV;
+
+    rcc2 = (sys_div << RCC2_SYSDIV_BITPOS); 
+
+    rcc  |= RCC_BYPASS | RCC_OFF;
+    rcc2 |= RCC2_BYPASS | RCC2_OFF;
+
+    RCC_CR = rcc;
+    RCC2_CR = rcc2;
+    __dmb();
```
Comment (karlp) [inline lib/lm3s/rcc.c]: Is the dmb here a lm3s documented requirement or something?  I admit, I've not read the lm3s ref mans at all, but it's unusual compared to other platforms.

File: lib/lm3s/rcc.c [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        ;
+
+    /* Stage 5: Disable bypass */
+    rcc  &= ~RCC_BYPASS;
+    rcc2 &= ~RCC2_BYPASS;
+    RCC_CR = rcc;
+    RCC2_CR = rcc2;
+    __dmb();
+    return 0;
+}
```
Comment (karlp) [inline lib/lm3s/rcc.c]: The entire style here of continually hard setting the RCC[2]_CR register is very odd to my (late night) eyes.  In most other platforms, you simply or in the bits you need to set, and check for them if necessary.  Continually modifying the rcc and rcc2 variables and continually hard assigning them to the RCC[2]_CR registers is... odd.

File: lib/lm3s/rcc.c [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+int rcc_qemu_init(void)
+{
+    int ret;
+    ret = rcc_lm3s_init(QEMU_SYSDIV_VAL, QEMU_PWMDIV_VAL, QEMU_XTAL_VAL);
+    return ret;
+}
```
Comment (karlp) [inline lib/lm3s/rcc.c]: related to my earlier comment on the defines, this kinda seems out of scope for core library.  Documented what values are good to use if you want to use qemu, but explicitly calling this a "qemu" init seems out of place

File: lib/lm3s/usart.c [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <libopencm3/lm3s/usart.h>
+
+
+void usart_set_baudrate(uint32_t usart, uint32_t baud)
```
Comment (karlp) [inline lib/lm3s/usart.c]: Please, if they're not implemented, just leave them out of the .h and leave them out of the .c file.  There's simply no reason for including empty stubs like this.

**Comment by karlp** [comment]
Generally looks good overall.
